course_flow/views.py

```diff
@@ -1027,3 +904,270 @@ def dialog_form_remove(request: HttpRequest) -> HttpResponse:
+                    node_type=model.node.node_type,
+                ),
+                rank=model.rank + 1,
+            )
+        else:
+            raise ValidationError
+
+    except ValidationError:
+        return JsonResponse({"action": "error"})
+
+    return JsonResponse({"action": "posted", "objectID": newmodel.id})
+
+
+"""
+Reorder methods
+"""
+
+# Insert a model via its throughmodel
+@require_POST
+@ajax_login_required
+@is_throughmodel_parent_owner
+def inserted_at(request: HttpRequest) -> HttpResponse:
+    object_id = json.loads(request.POST.get("objectID"))
+    object_type = json.loads(request.POST.get("objectType"))
+    parent_id = json.loads(request.POST.get("parentID"))
+    new_position = json.loads(request.POST.get("newPosition"))
+    new_parent_id = json.loads(request.POST.get("newParentID"))
+    new_column_id = json.loads(request.POST.get("newColumnID"))
+    try:
+        model_type = get_model_from_str(object_type)
+        model = model_type.objects.get(id=object_id)
+        old_position = model.rank
+
+        parentType = get_parent_model_str(object_type)
+
+        parent = get_model_from_str(parentType).objects.get(id=parent_id)
+
+        if not new_parent_id is None:
+            new_parent = get_model_from_str(parentType).objects.get(
+                id=new_parent_id
+            )
+        else:
+            new_parent = parent
+        new_parent_count = model_type.objects.filter(
+            **{parentType: new_parent}
+        ).count()
+        if new_position < 0:
+            new_position = 0
+        if new_position > new_parent_count:
+            new_position = new_parent_count - 1
+        delta = new_position - old_position
+
+        if parent.id == new_parent.id:
+            if delta != 0:
+                sign = int(math.copysign(1, delta))
+                for out_of_order_link in model_type.objects.filter(
+                    rank__gte=min(old_position + 1, new_position),
+                    rank__lte=max(new_position, old_position - 1),
+                    **{parentType: parent}
+                ):
+                    out_of_order_link.rank -= sign
+                    out_of_order_link.save()
+                model.rank = new_position
+                model.save()
+        elif parent.id != new_parent.id:
+            if hasattr(parent, "get_subclass"):
+                if (
+                    parent.get_subclass().author
+                    != new_parent.get_subclass().author
+                ):
+                    raise ValidationError
+            else:
+                if parent.author != new_parent.author:
+                    raise ValidationError
+            for out_of_order_link in model_type.objects.filter(
+                rank__gt=old_position, **{parentType: parent}
+            ):
+                out_of_order_link.rank -= 1
+                out_of_order_link.save()
+            for out_of_order_link in model_type.objects.filter(
```

Comment:
This type of stuff needs to be in models.py; the code should be made into a signal.